feat(http): Add Prefer, Preferrence-Applied headers (#37659)
mdn/content: files/en-us/web/http/reference/headers/prefer/index.md

---
title: Prefer
slug: Web/HTTP/Reference/Headers/Prefer
page-type: http-header
spec-urls: https://www.rfc-editor.org/rfc/rfc7240#section-2
---

{{HTTPSidebar}}

The HTTP **`Prefer`** header allows clients to indicate preferences for specific server behaviors during request processing.

> [!NOTE]
> Browsers have no handling for the `Prefer` and {{HTTPHeader("Preference-Applied")}} headers: they are used in custom, implementation-specific clients.
> Ensure both client and server support this header before relying on it in production.
>
> Servers should silently ignore preferences that they do not support, as though the header were not present.

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>
        {{Glossary("Request header")}}
      </td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>No</td>
    </tr>
  </tbody>
</table>

## Syntax

```http
Prefer: <preference>
```

## Directives

- `respond-async`

>   - : The client prefers asynchronous processing.

Reviewer: Should we outline anything here or leave it to the example, I tend to like a little more

```suggestion
  - : The client prefers asynchronous processing.
    For example, the server might respond with {{httpstatus("202", "202 (Accepted)")}} response indicated that the request has been accepted, along with the {{httpheader("Location")}} header that has a URL that the client can use to monitor the state of the processing.

```

Reviewer 2: Yeah this is good 👍🏻

The HTTP **`Prefer`** header allows clients to indicate preferences for specific server behaviors during request processing.

> [!NOTE]
> Browsers have no handling for the `Prefer` and {{HTTPHeader("Preference-Applied")}} headers: they are used in custom, implementation-specific clients.
> Ensure both client and server support this header before relying on it in production.
>
> The `Prefer` header does not cause the server to return an error if it does not support or apply the specified preferences.
> Instead, the server processes the request as if the header was not present.

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>
        {{Glossary("Request header")}}
      </td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>No</td>
    </tr>
  </tbody>
</table>

## Syntax

```http
Prefer: <preference>
```

## Directives

- `respond-async`
  - : The client prefers asynchronous processing.
- `return=minimal`
  - : Requests that the server return minimal content (a headers-only response).
- `return=representation`
  - : Requests a full resource representation in the response.
- `wait=<seconds>`

>   - : The maximum time the client expects it will take for the server to process the request once it's been received.
>     For example, if `respond-async` preference is also provided, the client requests that the service respond asynchronously after the specified length of time.
>     Otherwise, a server may interpret the wait time as a request to timeout after the specified period of time.

Reviewer: For example doesn't feel quite right here. 
- I think with `respond-async` this means "time in which I want my HTTP response" not "time that the whole thing will take.
- I think for sync responses we're saying "this is a timeout", but what happens then - the server still responds with something presumably.

If so, then in either case it is the time until a response is expected. So something like ...

```suggestion
  - : The time within which the client expects the server to take provide a response.
    If `respond-async` preference is also provided this is the time at which the initial response is provided (not completion of processing).
    For synchronous processing, a server may should timeout and respond if it unable to process the request in this timeframe.
```

Reviewer 2: I think the original is a little closer: https://datatracker.ietf.org/doc/html/rfc7240#section-4.3

>The "wait" preference can be used to establish an upper bound on the
   length of time, in seconds, the client expects it will take the
   server to process the request once it has been received.  In the case
   that generating a response will take longer than the time specified,
   the server, or proxy, can choose to utilize an asynchronous
   processing model by returning -- for example -- a 202 (Accepted)
   response.
   
This is probably the most relevant section:

>For example, a server receiving the following request might choose to
>   respond asynchronously if processing the request will take longer
>   than 10 seconds:
>
>     POST /collection HTTP/1.1
>     Host: example.org
>     Content-Type: text/plain
>     Prefer: respond-async, wait=10
>
>     {Data}

I interpret this as:

* `Prefer: respond-async` <- client can technically handle async responses
* `Prefer: wait=10` <- client will wait 10 seconds for a response, otherwise timeout
* `Prefer: respond-async, wait=10` <- respond asynchronously if processing the request will take longer than 10 seconds

So we could modify your suggestion to be:

```
The time within which the client expects the server to provide a response, from the point at which the request was received.
If the `respond-async` preference is also provided, the server should respond asynchronously if processing the request will exceed the wait time.
Otherwise, a server may consider `wait` the time it will take to send a synchronous response before the client times out.
```

Reviewer: That's helpful. So FMI if a client times out, what does it do with the response? What should  server send if it knows the client is timing out. That might affect what we write. See what I mean below with Xxxxx.

```suggestion
  - : The time within which the client expects the server to provide a response, from the point at which the request was received.
    If the `respond-async` preference is also provided, the server should respond asynchronously if processing the request will exceed the wait time.
    Otherwise, the server should consider that the client will timeout after the `wait` time and Xxxxx
```

Reviewer 2: This is great. Only problem is the "and Xxx" is not clearly speccd, so this depends on server implementation. Probably 408 https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/408 but the spec doesn't even go that far / make any assumptions what happens here. I think we're good with everything up to " and Xxxxx".

Reviewer: It's not ideal, but if that is the reality. I've decided to add a note that the behaviour is server dependent as this is useful.

```suggestion
  - : The time within which the client expects the server to provide a response, from the point at which the request was received.
    If the `respond-async` preference is also provided, the server should respond asynchronously if processing the request will exceed the wait time.
    Otherwise, the server should consider that the client will timeout after the `wait` time (response behavior depends on server implementation).
```

The HTTP **`Prefer`** header allows clients to indicate preferences for specific server behaviors during request processing.

> [!NOTE]
> Browsers have no handling for the `Prefer` and {{HTTPHeader("Preference-Applied")}} headers: they are used in custom, implementation-specific clients.
> Ensure both client and server support this header before relying on it in production.
>

> > The `Prefer` header does not cause the server to return an error if it does not support or apply the specified preferences.
> > Instead, the server processes the request as if the header was not present.

Reviewer: 1. Servers do whatever they like, so this is "should do" not "does do".
2. Perhaps simpler to say

```suggestion
> Servers should silently ignore preferences that they do not support, as though the header were not present.
```
